text/0000-sigstore-integration.md

```diff
@@ -0,0 +1,387 @@
+
+1. Authenticate the publisher using OIDC to retrieve a token
+1. Generate the ephemeral signing key and certificate request
+1. Generate a certificate signing request and pass with a token to Sigstore Fulcio.
+1. Sign the generated .crate file using the private key
+1. Attach signature, identity and certificate to crates.io publish request
+1. Publish an entry to the Rekor log with the .crate digest and certificate public key
+
+Most of the above is already implemented in https://github.com/sigstore/sigstore-rs.
+
+### crates.io flow
+
+The crates.io sub-flow on publish is described below:
+
+![crates.io flow](https://raw.githubusercontent.com/lulf/rfc-resources/main/sigstore/cargo_sigstore-crates.io.drawio.png)
+
+The following components need changing for crates.io:
+
+1. Additional columns must be added to the PostgreSQL database
+1. HTTP API must handle a new version of NewCratePublish request with the signature data
+1. The HTTP handler talks to Rekor and verifies that the signature is valid and that owner matches certificate
+
+## Cargo verify flow
+
+The cargo flow during verification is shown below:
+
+![verify flow](https://raw.githubusercontent.com/lulf/rfc-resources/main/sigstore/cargo_sigstore-verify.drawio.png)
+
+The following changes must be made to cargo when building/verifying a crate:
+
+1. Retrieve the crate contents, signature and certificate.
+1. Verify that the crate contents is signed by the signature.
+1. Verify that the certificate and signature is present in the transparency log.
+
+### Offline flow
+
+The offline verification flow is shown below
+
+![verify flow](https://raw.githubusercontent.com/lulf/rfc-resources/main/sigstore/cargo_sigstore-verify-offline.drawio.png)
+
+## Note on cargo dependencies
+
+The interaction with Sigstore services is implemented in the `sigstore-rs` crate, which has the following sub-dependencies:
+
+* tough: used to fetch fulcio public key and rekor certificates from Sigstore's TUF repository. 
+* openid-connect: this is used to generate keyless signatures via Rust
+* oauth2: this is a transitive dependency of openid-connect
+
+These crate use `reqwest` and some async Rust. To handle this, we can either:
+
+* Add the necessary dependencies to cargo with the goal of replacing the curl with reqwest to avoid multiple HTTP clients in use
+* Use the lower level APIs in sigstore-rs and implement the HTTP interaction using curl which is used by cargo today. 
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Cargo may be slower if signing and verification is enabled by default. To start, it should be made opt-in for a period so that maintainers and users can get to know the tool and we can learn more about the reliability.
+
+Reliance on the public [Sigstore uptime](https://www.chainguard.dev/unchained/sigstore-is-generally-available) of 99.5% availability goal may affect users signing and verifying crates, as well as crates.io verifying signature identities upon publish.
```

Comment:
If I don't misunderstand, the proposal also effectively relies on GitHub;s uptime as an OpenID Connect provider.

Reply:
crates.io already requires GitHub for auth, so, in a sense it's already burdened by that requirement. Plus, the entire project is strongly tethered to GH infrastructure, so I don't think too many people will object, even though I personally dislike it.

Reply:
We considered GitHub an existing piece of the infrastructure for crates.io due to the authentication process, which is the same OAuth flow used by Sigstore for generating the signing keys.

Reply:
Note that nothing right now prevents adding more authentication providers to crates.io, it's just that nobody implemented them yet. There is no intention of only ever supporting GitHub.

Reply:
Thanks, for bringing that up. I want to clarify that there is nothing about Sigstore that _requires_ GitHub - it can be used with other identity providers, and in fact the public instance supports multiple providers https://docs.sigstore.dev/fulcio/oidc-in-fulcio/#supported-oidc-token-issuers , and allows crates.io to 'run their own' in the future.

Reply:
(I'm a maintainer on Sigstore) Also worth noting we've been working on support for additional CI platforms like GitLab or Buildkite.

You can authenticate as a user interactively through various providers (Google, GitHub, Microsoft, and we can support more too), authenticate as a service account from Kubernetes, or authenticate via your preferred CI platform. GitHub Actions is the most fleshed out currently.